### PR TITLE
Migrate most of MixingModule and PreMixingModule to EventSetup consumes

### PR DESCRIPTION
**Pull request by makortel**
#### PR description:

This PR is part of #31061. In addition of plain migration to EventSetup consumes I did some refactoring related to getting the EventSetup products (like moving the `get()` calls from beginRun/beginLumi to Event, and passing around const references to ES products instead of `edm::ESHandle`). It has two caveats (which make the "most of") that need to be discussed/agreed:

I noticed the `MixingModule` (and `BMixingModule` base class, so affects all mixing modules) have the capability of updating the mixing parameters from the EventSetup, see e.g.
https://github.com/cms-sw/cmssw/blob/c51856dddbcd16b2f68d98df921df9d083abcbb6/SimGeneral/MixingModule/plugins/MixingModule.cc#L290-L301
https://github.com/cms-sw/cmssw/blob/c51856dddbcd16b2f68d98df921df9d083abcbb6/Mixing/Base/src/PileUp.cc#L235-L240
For the EventSetup consumes that alone is fine, but the pixel digitizer uses the bunchspacing to decide which of the two ES products it gets
https://github.com/cms-sw/cmssw/blob/c51856dddbcd16b2f68d98df921df9d083abcbb6/SimTracker/SiPixelDigitizer/plugins/SiPixelDigitizerAlgorithm.cc#L587-L595
In principle it could be possible for the product-to-be-got to change in the IOV boundaries of `MixingRcd` (if this mode of operation would be enabled), on the other hand updates after the first event are, in fact, already made impossible by
https://github.com/cms-sw/cmssw/blob/c51856dddbcd16b2f68d98df921df9d083abcbb6/SimTracker/SiPixelDigitizer/plugins/SiPixelDigitizer.cc#L260-L264
Nevertheless, it could be that the mixing parameters would be read from an EventSetup object instead of configuration. If this mode of operation needs to be supported, we need to figure out some way to get only (right) one of the `SiPixelDynamicInefficiency` products instead of both as in this PR. Possibilities include an ESProducer mayConsume-type feature, or abstracting the choice between the products into a new ESProducer. Anyway, the big question is **if MixingModule infrastructure has to continue to support changing mixing parameters based on `MixingModuleConfig` EventSetup product?**

[Answer below](https://github.com/cms-sw/cmssw/pull/31697#issuecomment-708095420): this mode should continue to be supported. To be done in a subsequent PR.

I did not migrate `DataMixingModule` (because it is not being used). Likely it becomes even more broken with this PR. **I hope this is acceptable.**

[Answer below](https://github.com/cms-sw/cmssw/pull/31697#issuecomment-708095420): Breaking `DataMixingModule` even more is acceptable.

#### PR validation:

Limited matrix runs.

### CODE REVIEW

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
Turns out this PR is incomplete (because I thought the behavior of https://github.com/cms-sw/cmssw/pull/31746 was already there). That does not affect my questions in the PR description though.

**Comment by mdhildreth** [comment]
@makortel I think breaking the DataMixer is ok for now.  As to the question of what sort of parameter modification should be supported: I'm guessing that the possibility of modification from EventSetup were introduced in order to support run-dependent mixing, which would have IOV boundaries that could, potentially, change all sorts of parameters. I don't think the Pixel dynamic inefficiency code cared about that, which is probably a bug.  We pretty much never generate 50ns bunch spacing and 25ns bunch spacing MC in the same job, however, so we never spotted it.  If it were me, I would keep the capability of on-the-fly modification.

**Comment by civanch** [comment]
+1

**Comment by perrotta** [comment]
@makortel is this still supposed to be a draft? Do you still consider this "incomplete", and do you plan to make it further compliant with #31746?

**Comment by makortel** [comment]
Yes, this is still supposed to be a draft. At least Mike's comment https://github.com/cms-sw/cmssw/pull/31697#issuecomment-708095420 needs to be addressed, and I'd like to make it at least further if not completely compliant with https://github.com/cms-sw/cmssw/pull/31746 (as of now I have very little means to actually check that :( )

**Comment by perrotta** [comment]
>  (as of now I have very little means to actually check that :( )

Running them together (in your private area)?

**Comment by makortel** [comment]
> > (as of now I have very little means to actually check that :( )
> 
> Running them together (in your private area)?

Won't help in until DIGI step becomes otherwise clean because another non-complying module could throw an exception first.

**Comment by makortel** [comment]
Thanks @mdhildreth for the feedback! I would probably go with
>  abstracting the choice between the products into a new ESProducer.

for the case of deciding the pixel inefficiency payload based on `MixingModuleConfig` (it will lead to a bit more set up if that ever will be used, but I think that should be acceptable)

**Comment by perrotta** [comment]
@makortel please let us know if there is any plan to go ahead with this PR

**Comment by perrotta** [comment]
-1
- (Only to keep this PR out of the reco radar until the planned development will get resumed)

**Comment by srimanob** [comment]
Hi @makortel 
What is the current plan for this migration? Thanks.

**Comment by makortel** [comment]
Rebased, and migrated all constructor calls to derived classes of `SiPixelGainCalibrationServiceBase` to pass the `edm::ConsumesCollector`. I consider this PR to be complete (given its size I think any remaining ES-consumes is best done in subsequent PRs). Concerning 

> **if MixingModule infrastructure has to continue to support changing mixing parameters based on MixingModuleConfig EventSetup product?**
> ...
> I would probably go with
> 
> > abstracting the choice between the products into a new ESProducer.
> 
> for the case of deciding the pixel inefficiency payload based on `MixingModuleConfig` (it will lead to a bit more set up if that ever will be used, but I think that should be acceptable)

I'd leave it to a subsequent PR (I have it essentially implemented) because it may require some iteration, whereas the changes in this PR should be rather straightforward.

**Comment by makortel** [comment]
@cmsbuild, please test

File: SimTracker/SiStripDigitizer/plugins/PreMixingSiStripWorker.cc [hunk 1]
```diff
@@ -65,6 +65,15 @@ class PreMixingSiStripWorker : public PreMixingWorker {
   edm::InputTag SiStripAPVPileInputTag_;
   std::string SistripAPVListDM_;  // output tag
 
+  edm::ESGetToken<TrackerGeometry, TrackerDigiGeometryRecord> const pDDToken_;
+  edm::ESGetToken<SiStripBadStrip, SiStripBadChannelRcd> const deadChannelToken_;
+  edm::ESGetToken<SiStripGain, SiStripGainSimRcd> const gainToken_;
+  edm::ESGetToken<SiStripNoises, SiStripNoisesRcd> const noiseToken_;
+  edm::ESGetToken<SiStripThreshold, SiStripThresholdRcd> const thresholdToken_;
+  //edm::ESGetToken<SiStripPedestals, SiStripPedestalsRcd> const pedestalToken_;
+  edm::ESGetToken<SiStripApvSimulationParameters, TrackerTopologyRcd> apvSimulationParametersToken_;
```
Comment (mmusich) [inline SimTracker/SiStripDigitizer/plugins/PreMixingSiStripWorker.cc:74]:
```suggestion
  edm::ESGetToken<SiStripApvSimulationParameters, SiStripApvSimulationParametersRcd> apvSimulationParametersToken_;
```
this explains the exceptions in the relval tests [[1](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2430a0/14336/runTheMatrixINPUT-results/250202.2_TTbar_13+TTbar_13INPUT+DIGIPRMXUP15APVSimu_PU25+RECOPRMXUP15_PU25+HARVESTUP15_PU25/step2_TTbar_13+TTbar_13INPUT+DIGIPRMXUP15APVSimu_PU25+RECOPRMXUP15_PU25+HARVESTUP15_PU25.log)], [[2](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2430a0/14336/runTheMatrixINPUT-results/250202.4_TTbar_13+TTbar_13INPUT+DIGIPRMXUP15APVSimu_PU25+RECOPRMXUP15_HIPM_PU25+HARVESTUP15_PU25/step2_TTbar_13+TTbar_13INPUT+DIGIPRMXUP15APVSimu_PU25+RECOPRMXUP15_HIPM_PU25+HARVESTUP15_PU25.log)]

```
----- Begin Fatal Exception 20-Apr-2021 01:57:53 CEST-----------------------
An exception of category 'NoProxyException' occurred while
   [0] Processing  Event run: 1 lumi: 37 event: 1804 stream: 0
   [1] Running path 'HLTAnalyzerEndpath'
   [2] Prefetching for module L1TRawToDigi/'hltGtStage2Digis'
   [3] Prefetching for module RawDataCollectorByLabel/'rawDataCollector'
   [4] Prefetching for module SiStripDigiToRawModule/'SiStripDigiToRaw'
   [5] Calling method for module PreMixingModule/'mixData'
Exception Message:
No data of type "SiStripApvSimulationParameters" with label "" in record "TrackerTopologyRcd"
 Please add an ESSource or ESProducer to your job which can deliver this data.
----- End Fatal Exception -------------------------------------------------
```

Comment (makortel) [reply SimTracker/SiStripDigitizer/plugins/PreMixingSiStripWorker.cc]: Thanks @mmusich! Fixed.

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by perrotta** [comment]
+1
- Mostly AlCa and Sim: for Reco there are only two classes in RecoLocalTrackers touched, with rather simple updates
- Jenkins tests pass, with no differences in output

**Comment by yuanchao** [comment]
+1
* migrated most of EventSetupRecord::get(ESHandle&) to EventSetup::getHandle(ESGetToken&); two in TODO explained
* WF matrix tests and unit tests all passed

**Comment by srimanob** [comment]
+Upgrade

**Comment by civanch** [comment]
+1

**Comment by silviodonato** [comment]
kind reminder @cms-sw/db-l2

**Comment by makortel** [comment]
@ggovi Could you please review and sign? Thanks.

**Comment by ggovi** [comment]
+1

**Comment by silviodonato** [comment]
+1

**Pull request merged**